doc/developer/design/20210413_source_sink_resource_sharing.md

```diff
@@ -0,0 +1,269 @@
+two separate Kafka consumers that read from the `top` topic:
+
+```
+       kafka_src/1                    kafka_src/2
+  +---------------------+        +---------------------+
+  | librdkafka consumer |        | librdkafka consumer |
+  +---------------------+        +---------------------+
+  |    Avro decoding    |        |    Avro decoding    |
+  +---------------------+        +---------------------+
+           |                               |
+         view1                           view2
+```
+
+In the steady state, this is quite wasteful. We're streaming the records from
+`top` over the network twice and decoding the Avro in each record twice. We'd
+prefer to run just one Kafka consumer that feeds both views:
+
+```
+               +---------------------+
+               | librdkafka consumer |
+               +---------------------+
+               |    Avro decoding    |
+               +---------------------+
+             /                        \
+            /                           \
+      kafka_src/1                     kafka_src/2
+          |                                |
+        view1                            view2
+```
+
+### Today's workaround
+
+If you know the complete set of columns you care about ahead of time, you can
+coax an efficient implementation out of Materialize today with an intermediate
+materialized view:
+
+```sql
+CREATE SOURCE kafka_src FROM KAFKA BROKER '...' TOPIC 'top' FORMAT AVRO ... ;
+CREATE MATERIALIZED VIEW kafka_view AS SELECT col1, col2 FROM kafka_src;
+CREATE VIEW view1 AS SELECT col1 FROM kafka_view;
+CREATE VIEW view2 AS SELECT col2 FROM kafka_view;
+```
+
+This creates only one librdkafka consumer by ensuring that `kafka_src` is
+only instantiated once.
+
+```
+       kafka_src/1
+  +---------------------+
+  | librdkafka consumer |
+  +---------------------+
+  |    Avro decoding    |
+  +---------------------+
+            |
+        kafka_view
+         /      \
+      view1    view2
+```
+
+The downside, of course, is that you must commit ahead of time to the columns
+you care about. What if, down the road, you realize you need a `view3` that
+extracts `col3`? You have two unappealing options:
+
+  1. Delete `kafka_view`, `view1`, and `view2`. Recreate `kafka_view` with
+    `col3` as well, then create `view1`, `view2`, and `view3` atop the new
+    `kafka_view`.
+
+     This is both annoying to type and will interrupt the availability of
+     `view1` and `view2`.
+
+  2. Create another instantiation of `kafka_src` by materializing `view3`
+     atop `kafka_src` directly.
+
+     This is ergonomic but results in a suboptimal steady state.
+
+### Dynamic resource sharing
+
+With dynamic resource sharing, you get all the ergonomics of creating views
+as you please, but a system that converges on the same efficience that you'd
+get if you'd pre-committed to those views. The key insight is that we can
+create a temporary backfill consumer when a new instance is instantiated:
+
```

Comment:
I'm definitely in support of dynamic resource sharing, but instead of having a "steady-state primary" and a "temporary backfill", I prefer having "old primary" and a "new primary". As in when "new_primary" catches up to "old primary", `view1` and `view2` will use the same consumer as `view3`, and "old primary" will be jettisoned.

A "old primary"/"new primary" would be more compatible with ongoing work to push LinearOperators/MFPs into sources (#2592). Currently, LinearOperators are only pushed into upserts and CSVs, but they are essential in upserts, so let's take upsert as an example.
Suppose:
* `view1` is defined as `select * from kafka_src where foo = 1`;
* `view2` is defined as `select * from kafka_src where bar = 2`;

In the "steady-state primary"/"temporary backfill" design, when `view2` is supposed to start using the "steady-state primary", we must somehow reconcile the views of the world in the two upsert operators. Then we must change the upsert operator on the steady-state primary to start filtering by `foo = 1 or bar = 2`.
```
          steady-state primary                    temporary backfill
         +---------------------+                 +---------------------+
         | librdkafka consumer |                 | librdkafka consumer |
         +---------------------+                 +---------------------+
         |    Upsert(foo=1)    |                 |   Upsert (bar=2)    |
         +---------------------+                 +---------------------+
       /                                                   |
      /                                                    |
     kafka_src/1                                  kafka_src/2
    |                                                      |
  view1                                                  view2
```

Whereas in the "old primary"/"new primary" model, when we create the new primary, we can `OR` the filters of the new view with the filters from the old view.
```
          old primary                                 new primary
         +---------------------+                 +------------------------+
         | librdkafka consumer |                 | librdkafka consumer    |
         +---------------------+                 +------------------------+
         |    Upsert(foo=1)    |                 | Upsert (foo=1 or bar=2)|
         +---------------------+                 +------------------------+
        /                                                  |
       /                                             Filter (bar=2)
      /                                                    |
     kafka_src/1                                  kafka_src/2
    |                                                      |
  view1                                                  view2
```

This way, when `view1` is moved over to the new primary, there is no need to change the existing upsert operator.

```
                  new primary
         +------------------------+
         | librdkafka consumer    |
         +------------------------+
         | Upsert (foo=1 or bar=2)|
         +------------------------+
        /                      |
  Filter(foo=1)       Filter (bar=2)
      /                        |
     kafka_src/1        kafka_src/2
    |                          |
  view1                    view2
```